### PR TITLE
Add r_buf_set_bytes_steal to eliminate copying

### PR DESCRIPTION
**Pull request by SchumBlubBlub**
https://github.com/radare/radare2/issues/6642
https://github.com/radare/radare2/pull/6719

reduce peak maximum mem during load. This introduces R_API function. 
For 1GB file:

before
before
After
after

steal too copies too many...

### CODE REVIEW

**Comment by Maijin** [comment]
Segfault -> https://api.travis-ci.org/jobs/199499649/log.txt?deansi=true

File: libr/util/buf.c [hunk 1]
```diff
@@ -287,6 +287,22 @@ R_API int r_buf_set_bytes(RBuffer *b, const ut8 *buf, ut64 length) {
...
+
+	b->length = length;
+	b->empty = 0;
+	return true;
+}
+
+R_API int r_buf_set_bytes_steal(RBuffer *b, const ut8 *buf, ut64 length, bool steal_ptr) {
+	if (length <= 0 || !buf) return false;
+	free (b->buf);
+	if(steal_ptr) {
```
Comment (radare) [inline libr/util/buf.c]: missing space before (

File: libr/util/buf.c [hunk 2]
```diff
@@ -287,6 +287,22 @@ R_API int r_buf_set_bytes(RBuffer *b, const ut8 *buf, ut64 length) {
...
+		b->buf = (ut8*) buf;
+	} else {
+		if (!(b->buf = malloc (length)))
```
Comment (radare) [inline libr/util/buf.c]: missing braces

File: libr/util/buf.c [hunk 3]
```diff
@@ -287,6 +287,22 @@ R_API int r_buf_set_bytes(RBuffer *b, const ut8 *buf, ut64 length) {
...
 		return false;
 	memmove (b->buf, buf, length);
+
+	b->length = length;
+	b->empty = 0;
+	return true;
+}
+
+R_API int r_buf_set_bytes_steal(RBuffer *b, const ut8 *buf, ut64 length, bool steal_ptr) {
+	if (length <= 0 || !buf) return false;
```
Comment (radare) [inline libr/util/buf.c]: missing braces

File: libr/util/buf.c [hunk 4]
```diff
@@ -287,6 +287,22 @@ R_API int r_buf_set_bytes(RBuffer *b, const ut8 *buf, ut64 length) {
 	if (!(b->buf = malloc (length)))
 		return false;
 	memmove (b->buf, buf, length);
+
+	b->length = length;
+	b->empty = 0;
+	return true;
+}
+
+R_API int r_buf_set_bytes_steal(RBuffer *b, const ut8 *buf, ut64 length, bool steal_ptr) {
```
Comment (radare) [inline libr/util/buf.c]: shouldnt this function just steal the pointer? whats the purpose of the bool parameter if there's another function that already does the set_byets() without stealing pointers?

Comment (SchumBlubBlub) [inline libr/util/buf.c]: I started by changing the sig for the original `r_buf_set_bytes`, and kept this when I decided it should be separate. wanted to know what you think. I'll remove it.

**Comment by radare** [comment]
lets wait for travis

**Comment by SchumBlubBlub** [comment]
haven't look at maijin's segfault yet, anyway.  Why is travis failing all PRs?

**Comment by radare** [comment]
Because master is red. Alvaro will fix tonight because its because of 2 tests that he is aware of that

**Comment by SchumBlubBlub** [comment]
r2r make all passes if I comment out the free I put there to prevent leaks, so is fixable. Tired now.

**Comment by radare** [comment]
so its a dblfree, also pls rebase, there's a conflict in bin/bin.c

File: libr/util/buf.c [hunk 5]
```diff
@@ -287,6 +287,16 @@ R_API int r_buf_set_bytes(RBuffer *b, const ut8 *buf, ut64 length) {
...
 		return false;
 	memmove (b->buf, buf, length);
+
+	b->length = length;
+	b->empty = 0;
+	return true;
+}
+
+R_API int r_buf_set_bytes_steal(RBuffer *b, const ut8 *buf, ut64 length) {
+	if (length <= 0 || !buf) return false;
```
Comment (radare) [inline libr/util/buf.c]: missing braces {}

Comment (SchumBlubBlub) [inline libr/util/buf.c]: Waiting for https://github.com/radare/radare2/pull/6749

**Comment by radare** [comment]
its still crashing

**Comment by SchumBlubBlub** [comment]
Haven't worked on that yet, just a rebase.

**Comment by SchumBlubBlub** [comment]
should green now.

**Comment by radare** [comment]
Thanks!

**Pull request merged**